allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -134,157 +100,85 @@ class DatasetReader(Registrable):
+        unless you implement handle the sharding within your `_read()` method, in which
+        case you should set `manual_multi_process_sharding` to `True`, just as with
+        `manual_distributed_sharding`.
 
-    """
+        See the note below about how to do this.
+
+    # Notes
+
+    The default mechanism for filtering out `Instance`s in the distributed or multi-process
+    `DataLoader` setting is not very efficient, since every worker would still need to
+    process every single `Instance` in your dataset.
+
+    This can be improved by manually handling the filtering / sharding within your `_read()`
+    method.
+
+    For example, if you were training using 2 GPUs and your `_read()` method reads a file
+    line-by-line, creating one `Instance` for each line, you could just check the node
+    rank within `_read()` and then throw away every other line starting at the line number
+    corresponding to the node rank.
+
+    The helper method `shard_iterable()` is there to make this easy for you.
+    You can wrap this around any iterable object in your `_read()` method, and it will
+    return an iterator that skips the right items based on the distributed training
+    or multi-process loading context. This method can always be called regardless
+    of whether or not you're actually using distributed training or multi-process loading.
+
+    Remember though that when you handle the sharding manually within `_read()`, you need
+    to let the `DatasetReader` know about this so that it doesn't do any additional
+    filtering. Therefore you need to ensure that both `self.manual_distributed_sharding` and
+    `self.manual_multi_process_sharding` are set to `True`.
+
+    If you call the helper method `shard_iterable()` without setting these to `True`,
+    you get an exception.
 
-    CACHE_FILE_LOCK_TIMEOUT: int = 10
-    """
-    The number of seconds to wait for the lock on a cache file to become available.
     """
 
     def __init__(
         self,
-        lazy: bool = False,
-        cache_directory: Optional[str] = None,
         max_instances: Optional[int] = None,
         manual_distributed_sharding: bool = False,
         manual_multi_process_sharding: bool = False,
     ) -> None:
-        self.lazy = lazy
+        # Do some validation.
+        if max_instances is not None and max_instances < 0:
+            raise ValueError("If specified, max_instances should be a positive int")
+
         self.max_instances = max_instances
-        self._cache_directory: Optional[Path] = None
-        if cache_directory:
-            self._cache_directory = Path(cache_directory)
-            os.makedirs(self._cache_directory, exist_ok=True)
         self.manual_distributed_sharding = manual_distributed_sharding
         self.manual_multi_process_sharding = manual_multi_process_sharding
+        self.__worker_info: Optional[WorkerInfo] = None
+        self.__dist_info: Optional[DistributedInfo] = None
```

Comment:
`dist` is pretty ambiguous; I'd recommend avoiding this abbreviation. (I also think that about the already-existing shortened import at the top of the file...)